### PR TITLE
Split nixos/nix-daemon.nix

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation

Goals:
 - make Nix more configurable. (personally I don't want to have anything `nix-channel` related on my system)
    - will follow up with actual additions / changes later
 - make modules easier to read
 - improve portability

For these goals it makes sense to clean up the "nix-daemon" module so that we have fewer responsibilities in each file.
This makes maintenance easier, including nix-darwin as well (after triaging their own module that was based on this one). After this, the modules that were factored out all happen to be very portable...

###### Description of changes

Move some definitions around.

`nix-daemon.nix` contained a lot of things that weren't the daemon. That's significantly reduced now (although admittedly not zero, and that's ok).
It does not make real changes to the resulting system configuration.
The only change that's close to "real" is splitting an activation script part into two parts: `nix` and `nix-channel`, rather than just `nix`. Other than that, the generated system derivations are unmodified.

Line count has increased due to typical per file overhead, but the smaller modules are easier to understand, so that's a big win.

Best reviewed per commit.

:speaker::notes:  _I don't like big modules and I don't think we should lie._ 

###### Things done

:heavy_check_mark: Double checked that I didn't accidentally remove a definition.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jsoo1** [comment]
I always get a bit lost looking for the nix modules. Instead of `nixos/modules/config` can they be in `nixos/modules/nix` or similar?

**Comment by roberth** [comment]
I've considered that, but then I didn't because that would be a significant change to the directory layout.
I'd like for something like RFC 140 to apply here - the one that puts all packages in an effectively flat directory, but that should be a community decision, not just me going wild :cowboy_hat_face: 
The locations I did end up picking are the ones that seem most applicable, and to be honest I think they do provide some useful structure. Not structure that couldn't be achieved otherwise, but welcome structure nonetheless.

**Comment by roberth** [comment]
I guess one thing I could do is add some `imports` so that the other modules pull in the one with `nix.conf`, if that helps a bit.

**Comment by jsoo1** [comment]
> I guess one thing I could do is add some `imports` so that the other modules pull in the one with `nix.conf`, if that helps a bit.

That would be nice, but I won't push too hard. I wouldn't block on the directory layout.

**Comment by roberth** [comment]
Ok, I did add two imports because they would be technically necessary if you were to cherry pick modules and evaluate them separately. All of the links are also in the top comment though, so that's the more useful place to look. I've only linked the relevant modules, but all are reachable in 2 hops when going via `nix.nix`.

**Comment by jsoo1** [comment]
Lovely! Thank you!

File: nixos/modules/config/nix.nix [hunk 1]
```diff
@@ -0,0 +1,379 @@
+/*
+  Manages /etc/nix.conf.
+
+  See also
+   - ./nix-channel.nix
+   - ./flakes.nix
+   - ./nix-remote-build.nix
```
Comment (SuperSandro2000) [inline nixos/modules/config/nix.nix:7]:
```suggestion
   - ./flakes.nix
   - ./nix-channel.nix
   - ./nix-remote-build.nix
```

File: nixos/modules/module-list.nix [hunk 2]
```diff
@@ -2,6 +2,7 @@
   ./config/appstream.nix
   ./config/console.nix
   ./config/debug-info.nix
+  ./config/flakes.nix
```
Comment (SuperSandro2000) [inline nixos/modules/module-list.nix:5]:
```suggestion
  ./config/nix-flakes.nix
```
all other things include nix so we might as well

File: nixos/modules/services/system/nix-daemon.nix [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  };
+
+  nixbldUsers = listToAttrs (map makeNixBuildUser (range 1 cfg.nrBuildUsers));
+
+in
+
+{
+  imports = [
+    ../../config/nix.nix
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/system/nix-daemon.nix:44]: isn't this already included in the modules list?

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/modules/services/system/nix-daemon.nix (hunk 3 above)
Comment (roberth) [reply nixos/modules/services/system/nix-daemon.nix:44]: It declares that you'd need a global nix configuration when you want to use the daemon. It's not technically necessary, but it is informative and potentially useful if someone wants to cherry pick modules.

File: nixos/modules/misc/version.nix [hunk 4]
```diff
@@ -140,13 +140,6 @@ in
       '';
     };
 
-    defaultChannel = mkOption {
-      internal = true;
-      type = types.str;
-      default = "https://nixos.org/channels/nixos-unstable";
-      description = lib.mdDoc "Default NixOS channel to which the root user is subscribed.";
-    };
```
Comment (Ericson2314) [inline nixos/modules/misc/version.nix:148]: Oh gawd, that was here?!

**Review by Ericson2314 — APPROVED** [review]
This is great! Only thing I might rename is `nix-remote-build` to -> `nix-machines` or something, to reflect the fact that it is only separate because a legacy split of the configuration. (e.g. in a future JSON Nix config world, I would expect those settings to be folded back in and configured the same way as everything else.)

**Comment by roberth** [comment]
> `nix-remote-build` to -> `nix-machines`

Let's just remove it at some point. I don't think we have a winning name for this anyway, as `machines` isn't very informative, and the options don't suggest a great alternative either.

**Comment by roberth** [comment]
Thanks for the reviews :heart:

File: nixos/modules/services/system/nix-daemon.nix (hunk 3 above)
Comment (SuperSandro2000) [reply nixos/modules/services/system/nix-daemon.nix:44]: > It's not technically necessary, but it is informative and potentially useful if someone wants to cherry pick modules.

AFAIK we don't do this anywhere else and removing/picking any module probably draws in all the time a long tail fo dependencies.

Comment (SuperSandro2000) [reply nixos/modules/services/system/nix-daemon.nix:44]: #242028

**Comment by roberth** [comment]
You may also like
 - https://github.com/NixOS/nixpkgs/pull/242098

**Pull request merged**